torch/csrc/utils/tensor_numpy.cpp

```diff
@@ -227,6 +230,10 @@ ScalarType numpy_dtype_to_aten(int dtype) {
       ((PyTypeObject*)pytype.get())->tp_name);
 }
 
+bool is_numpy_int(PyObject* obj) {
```

Comment:
For np.bool:
- `is_numpy_int` returns false
- `is_numpy_scalar` returns true

Is that correct?

Reply:
yes

Reply:
I stepped through this in gdb. `PyArray_IsIntegerScalar` returns False for np.bool, so we don't need the extra check for bool. The relevant source code is here: https://github.com/numpy/numpy/blob/4bc7eda1f0ff87a9e6eaace60ef66b6c1b9658a1/numpy/core/include/numpy/ndarrayobject.h#L61-L63 but it's a little hard to read.

In addition, we could clean up `is_numpy_scalar` by using `is_numpy_int`.

Reply:
If I'm understanding correctly, that seems to imply that all of the existing code like `return (PyLong_Check(obj) || PyInt_Check(obj)) && !PyBool_Check(obj);` is also incorrect? 

PyArray_IsIntegerScalar etc appears to rely on the same internal interpretation of looking for Py_TPFLAGS_LONG_SUBCLASS on the type object's tp_flags, which seems isn't set for bool types.

Reply:
> If I'm understanding correctly, that seems to imply that all of the existing code like `return (PyLong_Check(obj) || PyInt_Check(obj)) && !PyBool_Check(obj);` is also incorrect?

No, the existing code is correct because PyLong_Check returns True for Python booleans.

`PyArray_IsIntegerScalar` returns False for np.bool, but True for PyBool, which is kinda fun. So given that, my question is, what should `is_numpy_int` actually check? What does it return for np.int64, PyLong, np.bool, PyBool, np.double, PyFloat?

Reply:
hmm... It seems PyLong_Check effectively just checks `((t)->tp_flags & Py_TPFLAGS_LONG_SUBCLASS) != 0`, where Py_TPFLAGS_LONG_SUBCLASS == 1 << 24. Is that not right?

It appears that np.bool and bool should behave the same?

```
In [14]: bool.__flags__ & (1 << 24)
Out[14]: 16777216
In [15]: np.bool.__flags__ & (1 << 24)
Out[15]: 16777216
```
What am I misunderstanding about PyLong_Check?

Reply:
I'm inferring the behavior of PyLong_Check from the documentation. I haven't actually run any code or read the implementation of PyLong_Check.

Based on the documentation:
- The documentation says "Return true if its argument is a PyLongObject or a subtype of PyLongObject.": https://docs.python.org/3/c-api/long.html#c.PyLong_Check.
- Then, in the documentation for boolean, "Booleans in Python are implemented as a subclass of integers." https://docs.python.org/2/c-api/bool.html

What happens empirically if you write a program that checks the value of `PyLong_check(Py_True)`?

Reply:
It looks like there's a difference between np.bool and np.bool_: https://joergdietrich.github.io/python-numpy-bool-types.html#Numpy-bool. `np.bool` is the same thing as Python bool, while `np.bool_` is what you get if you create a bool array in numpy

Reply:
interesting... 
```
In [10]: np.array([True]).dtype.type
Out[10]: numpy.bool_
In [11]: np.array([True]).dtype.type.__flags__ & (1 << 24 )
Out[11]: 0
```